#   casaba.rb - description
#
# Copyright:: Copyright (C) 2012
#     Raoul Bonnal <@bioruby.org>
# License:: The Ruby License
#
#
# configureQseqToFastq.pl 
# Usage:
#     configureQseqToFastq.pl --input-dir DIR [options] | --help

#   Summary:
#     Create a makefile to convert a directory of qseq files to a directory
#     tree of compressed fastq files following CASAVA 1.8 filename and
#     directory structure conventions. If detected, configuration data used by
#     GERALD are also transfered to the output directory.

#     This script will not configure demultiplexing. The input directory must
#     contain qseq files which are either non-demultiplexed or already
#     demultiplexed by another utility.

#   Options:
#     --input-dir DIRECTORY
#         Path to qseq directory (no default)

#     --output-dir DIRECTORY
#         Path to root of CASAVA 1.8 unaligned directory structure. Directory
#         will be created if it does not exist (default:
#         '<input-dir>/QseqToFastq/Unaligned')

#     --fastq-cluster-count INTEGER
#         Maximum number of fastq records per fastq file (default: 4000000)

#     --config-file FILENAME
#         Specify the Bustard config file to be copied to the fastq directory
#         (default: '<input-dir>/config.xml')

#     --summary-file FILENAME
#         Specify the Bustard summary file to be copied to the fastq directory
#         (default: '<input-dir>/BustardSummary.xml')

#     --flowcell-id STRING
#         Use the specified string as the flowcell id. (default value is
#         parsed from the config-file)

module Bio
  module Ngs
    class CASAVA    
        class Bclqseq

          include Bio::Command::Wrapper

          set_program Bio::Ngs::Utils.binary("setupBclToQseq.py")

          add_option "base-calls-directory", :type => :string, :aliases =>'-b', :desc => 'Full path to the Base Calls directory generated by the instrument'
          add_option "in-place",:type => :boolean, :desc => 'Allow in-place conversion (in the Base Calls directory)'
          add_option "output-directory", :type => :string, :aliases => '-o', :desc => 'Full path to the output directory, if not in --in-place mode'
          add_option "intensities-directory", :type => :string, :aliases => '-i', :desc => 'Full path to the directory containing the intensities, if different from the parent of the Base Calls directory'
          add_option "filter-directory", :type => :string , :aliases => '-f', :desc => 'Full path to the filter directory, if different from the Base Calls directory'
          add_option "positions-directory", :type => :string, :aliases => '-p', :desc => 'Full path to the positions directory, if different from the Intensities directory'
          add_option "positions-format", :type => :string, :aliases => '-P', :desc => 'Positions file format (supported formats are: {\'_pos.txt\',\'.locs\',\'.clocs\'})'
          add_option "no-eamss", :type => :boolean, :desc => 'Do not apply the EAMSS quality masking'
          add_option "overwrite", :type => :boolean, :desc => 'Overwrite the content of the output directory if it already exists'
          add_option "GERALD", :type => :string, :desc => 'Path to the GERALD config file'
          add_option "silent", :type => :boolean, :aliases => '-S', :desc => 'Do not produce any information or warnings'
          add_option "quiet", :type => :boolean, :aliases => '-Q', :desc => 'Produce only warnings. Skips general information'
          add_option "verbose", :type => :boolean, :aliases => '-V', :desc => 'Produce additional debug information'
          add_option "log-file", :type => :string, :aliases => '-L', :desc => 'Full path to the log file'
          add_option "include-controls", :type => :boolean, :aliases => '-c', :desc => 'Include controls information in Passing Filter column'
          add_option "ignore-missing-bcl", :type => :boolean, :desc => 'Assume missing *.bcl files correspond to a missed base call (i.e., \'.\')'
          add_option "ignore-missin-stats", :type => :boolean, :desc => 'Fill in with zeros when *.stats files are missing'

      end #Bcl2seq

      class ConfigBclFastq
        include Bio::Command::Wrapper

        set_program Bio::Ngs::Utils.binary("configureBclToFastq.pl")

        # Usage:
        #     configureBclToFastq.pl
        #                 [--adapter-sequence adapter_fasta_file_path]
        #                 [--use-bases-mask mask] [--no-eamss] [--with-failed-reads]
        #                 [--input-dir base_calls_dir]
        #                 [--intensities-dir intensities_dir]
        #                 [--positions-dir positions_dir]
        #                 [--positions-format .locs .clocs or _pos.txt]
        #                 [--filter-dir filter_dir] [--output-dir output_dir]
        #                 [--sample-sheet sample_sheet]
        #                 [--mismatches num_of_mismatches per barcode component]
        #                 [--fastq-cluster-count cluster_count]
        #                 [--ignore-missing-stats] [--ignore-missing-bcl]
        #                 [--ignore-missing-control] [--flowcell-id flow_cell_id]
        #                 [--tiles tile_selection]

        #     configureBclToFastq.pl
        #                 --help or --man

        # Options and Arguments:
        #     --adapter-sequence adapter_fasta_file_path
        #             Path to a multi-contig fasta file. If only one sequence is
        #             specified within the fasta file, the same sequence would be used
        #             for all reads. If multiple sequences are specified, the first
        #             sequence would be used for the first read, the second sequence
        #             for the second read.

        #     --use-bases-mask mask[[,mask]...]
        #             Conversion mask characters:

        #               - Y or y: use
        #               - N or n: discard
        #               - I or i: use for indexing

        #             If not given, the mask will be guessed from the RunInfo.xml file
        #             in the run folder.

        #             For instance, in a 2x76 indexed paired end run, the mask
        #             Y76,I6n,y75n means: "use all 76 bases from the first end,
        #             discard the last base of the indexing read, and use only the
        #             first 75 bases of the second end".

        #     --no-eamss Disable the masking of the quality values with EAMSS.

        #     --with-failed-reads Include failed reads into the FASTQ files (by
        #     default, only reads passing filter are included). default, only reads passing filter are included).

        #     --input-dir base_calls_dir
        #             Path to a valid BaseCalls directory (defaults to current dir)

        #     --intensities-dir intensities_dir
        #             Path to a valid Intensities directory (defaults to parent of
        #             base_calls_dir)

        #     --positions-dir positions_dir
        #             Path to a directory containing positions files. (defaults
        #             depending on RTA version)

        #     --positions-format .locs .clocs or _pos.txt
        #             Format of the input cluster positions information. (defaults to
        #             .clocs)

        #     --filter-dir filter_dir
        #             Path to a directory containing filter files. (defaults depending
        #             on RTA version)

        #     --output-dir output_dir
        #             Path to the demultiplexed output (defaults to
        #             base_calls_dir/../../../Unaligned)

        #     --sample-sheet sample_sheet
        #             Path to SampleSheet.csv (defaults to
        #             base_call_dir/SampleSheet.csv)

        #     --mismatches num_of_mismatches per barcode component
        #             Comma-separated list of mismatches allowed for each barcode
        #             component. Either 0 or 1 (defaults to 0)

        #     --fastq-cluster-count cluster_count
        #             Maximum number of fastq records per fastq file (default:
        #             4000000).

        #     --ignore-missing-stats
        #             Fill in with zeros when *.stats files are missing

        #     --ignore-missing-bcl
        #             Interpret missing *.bcl files as no call

        #     --ignore-missing-control
        #             Interpret missing control files as not-set control bits

        #     --tiles regex[[,regex]...]
        #             Comma-separated list of regular expressions to select only a
        #             subset of the tiles available in the flow-cell.

        #              - to select all the tiles ending with "5" in all lanes: --tiles [0-9][0-9][0-9]5
        #              - to select tile 2 in lane 1 and all the tiles in the other lanes: --tiles s_1_0002,s_[2-8]

        #     --flowcell-id flow_cell_id
        #             Use the specified string as the flowcell id. (default value is
        #             parsed from the config-file)

        #     --help Print a brief help message and exit.

        #     --man View this help formatted in "man" style.

        #     Running the small test data set included with CASAVA
        #             Note that with the example below, if the SampleSheet.csv file is
        #             not found in the BaseCalls folder, the data is assumed to be
        #             non-multiplexed. Use "--sample-sheet" option to override default
        #             SampleSheet.csv location.

        #              /opt/CASAVA_v1.8.2/bin/configureBclToFastq.pl \
        #              --output-dir ./Unaligned \
        #              --input-dir /opt/CASAVA_v1.8.2/share/CASAVA-1.8.2/examples/Validation/Default/Data/Intensities/BaseCalls

        #               make -C Unaligned  # use make -j <parallel jobs> to speedup the conversion


          add_option "adapter-sequence", :type => :string, :desc => 'Path to a multi-contig fasta file. If only one sequence is specified within the fasta file, the same sequence would be used for all reads. If multiple sequences are specified, the first sequence would be used for the first read, the second sequence for the second read.'
          add_option "use-bases-mask", :type => :string, :desc => 'If not given, the mask will be guessed from the RunInfo.xml file in the run folder.'
          add_option "no-eamss", :type => :boolean, :desc => 'Disable the masking of the quality values with EAMSS.'
          add_option "with-failed-reads", :type => :boolean, :desc => 'Include failed reads into the FASTQ files (by default, only reads passing filter are included).'
          add_option "input-dir", :type => :string, :aliases =>'-i', :desc => 'Path to a valid BaseCalls directory (defaults to current dir)'
          add_option "intensities-dir", :type => :string, :desc => 'Path to a valid Intensities directory (defaults to parent of base_calls_dir)'
          add_option "positions-dir", :tpe => :string, :desc => 'Path to a directory containing positions files. (defaults depending on RTA version)'
          add_option "positions-format", :type => :string, :desc => 'Format of the input cluster positions information. (defaults to .clocs)'
          add_option "filter_dir", :type => :string, :desc => 'Path to a directory containing filter files. (defaults depending on RTA version)'
          add_option "output-dir",:type => :string, :aliases => '-o', :desc => 'Path to the demultiplexed output (defaults to base_calls_dir/../../../Unaligned)'
          add_option "sample-sheet", :type => :string, :desc => 'Path to SampleSheet.csv (defaults to base_call_dir/SampleSheet.csv)'
          add_option "mismatches", :type => :numeric, :desc => 'Comma-separated list of mismatches allowed for each barcode component. Either 0 or 1 (defaults to 0)', :defaults=> 1
          add_option "fastq-cluster-count", :type => :numeric, :desc => 'Maximum number of fastq records per fastq file (default: 4000000).'
          add_option "ignore-missin-stats", :type => :boolean, :desc => 'Fill in with zeros when *.stats files are missing'
          add_option "ignore-missing-bcl", :type => :boolean, :desc => 'Interpret missing *.bcl files as no call'
          add_option "ignore-missing-control", :type => :boolean, :desc => 'Interpret missing control files as not-set control bits'
          add_option "tiles", :type => :string, :desc => 'Comma-separated list of regular expressions to select only a subset of the tiles available in the flow-cell.'
          add_option "flowcell-id", :type => :string , :desc => 'Use the specified string as the flowcell id. (default value is parsed from the config-file)'
      end #ConfigBclFastq

      class ConfigQseqFastq
        include Bio::Command::Wrapper

        set_program Bio::Ngs::Utils.binary("configureQseqToFastq.pl")

        # Usage:
        #     configureQseqToFastq.pl --input-dir DIR [options] | --help

        #   Summary:
        #     Create a makefile to convert a directory of qseq files to a directory
        #     tree of compressed fastq files following CASAVA 1.8 filename and
        #     directory structure conventions. If detected, configuration data used by
        #     GERALD are also transfered to the output directory.

        #     This script will not configure demultiplexing. The input directory must
        #     contain qseq files which are either non-demultiplexed or already
        #     demultiplexed by another utility.

        #   Options:
        #     --input-dir DIRECTORY
        #         Path to qseq directory (no default)

        #     --output-dir DIRECTORY
        #         Path to root of CASAVA 1.8 unaligned directory structure. Directory
        #         will be created if it does not exist (default:
        #         '<input-dir>/QseqToFastq/Unaligned')

        #     --fastq-cluster-count INTEGER
        #         Maximum number of fastq records per fastq file (default: 4000000)

        #     --config-file FILENAME
        #         Specify the Bustard config file to be copied to the fastq directory
        #         (default: '<input-dir>/config.xml')

        #     --summary-file FILENAME
        #         Specify the Bustard summary file to be copied to the fastq directory
        #         (default: '<input-dir>/BustardSummary.xml')

        #     --flowcell-id STRING
        #         Use the specified string as the flowcell id. (default value is
        #         parsed from the config-file)  
          add_option "input-dir", :type => :string, :aliases =>'-i', :desc => 'Path to qseq directory (no default)'
          add_option "output-dir",:type => :string, :aliases => '-o', :desc => 'Path to root of CASAVA 1.8 unaligned directory structure. Directory will be created if it does not exist (default:\'<input-dir>/QseqToFastq/Unaligned\')'
          add_option "fastq-cluster-count", :type => :numeric, :aliases => '-c', :desc => 'Maximum number of fastq records per fastq file (default: 4000000)'
          add_option "config-file", :type => :string, :aliases => '-f', :desc => 'Specify the Bustard config file to be copied to the fastq directory (default: \'<input-dir>/config.xml\')'
          add_option "summary-file", :type => :string , :aliases => '-s', :desc => 'Specify the Bustard summary file to be copied to the fastq directory (default: \'<input-dir>/BustardSummary.xml\')'
          add_option "flowcell-id", :type => :string, :aliases => '-d', :desc => 'Use the specified string as the flowcell id. (default value is parsed from the config-file)'
      end #ConfigQseqFastq

    end #CASAVA
  end #Ngs
end #Bio 